rfc/006-backfill-blocks.md

```diff
@@ -0,0 +1,155 @@
+# RFC 006: Backfill blocks
+
+## Changelog
+
+- 2020-11-25: Initial draft.
+
+## Author(s)
+
+- Callum Waters (@cmwaters)
+
+## Context
+
+Two new features: [Block pruning](https://github.com/tendermint/tendermint/issues/3652)
+and [State sync](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-042-state-sync.md)
+meant nodes no longer needed a complete history of the blockchain. This
+introduced some challenges of its own which were covered and subsequently
+tackled with [RFC-001](https://github.com/tendermint/spec/blob/master/rfc/001-block-retention.md).
+The RFC allowed applications to set a block retention height; an upper bound on
+what blocks would be pruned. However nodes who state sync past this upper bound
+(which is necessary as snapshots must be saved within the trusting period for
+the assisting light client to verify) have no means of backfilling the blocks
+to meet the retention limit. This could be a problem as nodes who state sync and
+then eventually switch to consensus may not have the block and validator
+history to verify evidence causing them to panic if they see 2/3 commit on what
+the node believes to be an invalid block.
+
+Having the capability to backfill prior blocks also allows nodes with
+truncated history to become archive nodes which may be beneficial for the
+network.                                            
+
+## Proposal
+
+A node, seeking to state sync, will find an adequate snapshot and offer it to
+the application with the following format:
+
+```golang
+type snapshot struct {
+ 	Height   uint64
+ 	Format   uint32
+ 	Chunks   uint32
+ 	Hash     []byte
+ 	Metadata []byte
+
+ 	trustedAppHash []byte // populated by light client
+}
+```
+
+The application will then be expected to parse information in the meta data
+regarding the backfill height and then return it in the response (alongside the
+app's decision):
+
+```proto
+message ResponseOfferSnapshot {
+  Result result = 1;
+  int64 backfill_height = 2;
+}
+```
```

Comment:
I think the backfill height should be part of the snapshot metadata, and thus given by the node that took the snapshot. Remember, the local application has no state (it is about to be restored from a snapshot), so it may not have  necessary information to calculate the backfill height. The application that took the snapshot does. E.g.:

```protobuf
message Snapshot {
  uint64 height          = 1;  // The height at which the snapshot was taken
  uint32 format          = 2;  // The application-specific snapshot format
  uint32 chunks          = 3;  // Number of chunks in the snapshot
  bytes  hash            = 4;  // Arbitrary snapshot hash, equal only if identical
  bytes  metadata        = 5;  // Arbitrary application metadata
  int64  backfill_height = 6;  // Height to backfill blocks from before starting application (inclusive)
}
```

Reply:
ok I can change that